ocean_provider/routes/decrypt.py

```diff
@@ -0,0 +1,172 @@
+#
+# Copyright 2021 Ocean Protocol Foundation
+# SPDX-License-Identifier: Apache-2.0
+#
+import logging
+import lzma
+import traceback
+from hashlib import sha256
+from typing import Optional
+
+from eth_typing.encoding import HexStr
+from flask import Response, request
+from flask_sieve import validate
+from ocean_provider.log import setup_logging
+from ocean_provider.requests_session import get_requests_session
+from ocean_provider.user_nonce import increment_nonce
+from ocean_provider.utils.basics import (
+    LocalFileAdapter,
+    get_config,
+    get_provider_wallet,
+    get_web3,
+)
+from ocean_provider.utils.data_nft import (
+    MetadataState,
+    get_metadata,
+    get_metadata_logs_from_tx_receipt,
+)
+from ocean_provider.utils.encryption import do_decrypt
+from ocean_provider.utils.error_responses import error_response, service_unavailable
+from ocean_provider.utils.util import get_request_data
+from ocean_provider.validation.provider_requests import DecryptRequest
+from web3.main import Web3
+
+from . import services
+
+setup_logging()
+provider_wallet = get_provider_wallet()
+requests_session = get_requests_session()
+requests_session.mount("file://", LocalFileAdapter())
+
+logger = logging.getLogger(__name__)
+
+
+@services.route("/decrypt", methods=["POST"])
+@validate(DecryptRequest)
+def decrypt():
+    data = get_request_data(request)
+    logger.info(f"decrypt endpoint called. {data}")
+
+    try:
+        return _decrypt(
+            decrypter_address=data.get("decrypterAddress"),
+            chain_id=data.get("chainId"),
+            transaction_id=data.get("transactionId"),
+            data_nft_address=data.get("dataNftAddress"),
+            encrypted_document=data.get("encryptedDocument"),
+            flags=data.get("flags"),
+            document_hash=data.get("documentHash"),
+        )
+    except Exception as e:
+        return service_unavailable(e, data, logger)
+
+
+def _decrypt(
+    decrypter_address: HexStr,
+    chain_id: int,
+    transaction_id: Optional[HexStr],
+    data_nft_address: Optional[HexStr],
+    encrypted_document: Optional[HexStr],
+    flags: Optional[int],
+    document_hash: Optional[HexStr],
+) -> Response:
+    increment_nonce(decrypter_address)
+
+    # Check if given chain_id matches Provider's chain_id
+    web3 = get_web3()
+    if web3.eth.chain_id != chain_id:
+        return error_response(f"Unsupported chain ID", 400)
+
+    # Check if decrypter is authorized
+    authorized_decrypters = get_config().authorized_decrypters
+    logger.info(f"authorized_decrypters = {authorized_decrypters}")
+    if authorized_decrypters and decrypter_address not in authorized_decrypters:
+        return error_response(f"Decrypter not authorized", 403)
+
+    if transaction_id:
```

Comment:
can we switch this? i.e. start with `if not transaction_id`, since the block is shorter. Return what there is to return, and then there is no need for the "else" and indent.

Reply:
The `else` is important in this case. Only one of the two blocks should be executed.  Nonetheless, I've inverted the conditional and shortened these blocks by extracting the logic to two helper methods.